File: tests/tcpdns2pcap.lua
#!/usr/bin/env dnsjit
-- Convert a raw DNS-TCP-framed stream (2-byte BE length prefix + DNS wire
-- message, repeated) on stdin into a PCAP file containing IPv6/UDP/DNS
-- packets on stdout.
--
-- Writes a classic (non-nanosecond) PCAP file directly (LINKTYPE_RAW=101,
-- raw IP packets, no link-layer header) using plain Lua string building,
-- since dnsjit's output.pcap module has no public constructor for
-- synthetic core_object_pcap_t objects.
--
-- Always uses ::1 on both ends, port 53 on both ends.
--
-- usage: dnsjit tcpdns2pcap.lua [spacing_us] < in.tcpdns > out.pcap

local bit = require("bit")

local spacing_us = tonumber(arg[2]) or 300000

local src_port = 53
local dst_port = 53

local IP6_LOOPBACK = string.rep("\0", 15) .. "\1" -- ::1

local function u16_be(v)
	return string.char(
		bit.band(bit.rshift(v, 8), 0xff),
		bit.band(v, 0xff))
end

local function u32_be(v)
	return string.char(
		bit.band(bit.rshift(v, 24), 0xff),
		bit.band(bit.rshift(v, 16), 0xff),
		bit.band(bit.rshift(v, 8), 0xff),
		bit.band(v, 0xff))
end

local function read_u16_be(str, pos)
	return bit.bor(bit.lshift(str:byte(pos), 8), str:byte(pos + 1))
end

local function build_packet(dns_wire)
	local udp_len = 8 + #dns_wire

	-- UDP checksum is left as 0 (disabled/ignored): dnsjit's packet-layer
	-- parser never validates it, it only reads ports/length to locate the
	-- DNS payload. See pcap/generate-const-qps.lua, which does the same.
	local udp_hdr = u16_be(src_port) .. u16_be(dst_port) .. u16_be(udp_len) .. u16_be(0)

	local ip6_hdr = string.char(0x60, 0x00, 0x00, 0x00) -- version=6, traffic class/flow label=0
		.. u16_be(udp_len) -- payload length
		.. string.char(17) -- next header = UDP
		.. string.char(64) -- hop limit
		.. IP6_LOOPBACK -- source address
		.. IP6_LOOPBACK -- dest address

	return ip6_hdr .. udp_hdr .. dns_wire
end

-- LINKTYPE_RAW per the tcpdump.org linktype registry (portable across
-- platforms) rather than the OS-local DLT_RAW (12 on Linux, 14 on OpenBSD).
-- See pcap/generate-const-qps.lua, which uses the same value.
local LINKTYPE_RAW = 101

-- Written fully big-endian, including the magic number itself. Readers
-- (libpcap, dnsjit's fpcap/mmpcap) detect the byte-swapped magic
-- (0xa1b2c3d4 on disk reads back as 0xd4c3b2a1 on our little-endian host)
-- and transparently byte-swap every other field on read -- this is the
-- standard "swapped-endian" pcap dialect, not a dnsjit extension.
local hdr = u32_be(0xa1b2c3d4) -- magic
	.. u16_be(2) -- version_major
	.. u16_be(4) -- version_minor
	.. u32_be(0) -- thiszone
	.. u32_be(0) -- sigfigs
	.. u32_be(65535) -- snaplen
	.. u32_be(LINKTYPE_RAW) -- network
io.stdout:write(hdr)

-- stream tcpdns messages one at a time: read length prefix, read that many
-- bytes, build+write the packet immediately -- no buffering of the whole
-- input stream or the parsed message list in memory.
local ts_sec = 0
local ts_usec = 0
local count = 0
while true do
	local len_bytes = io.stdin:read(2)
	if not len_bytes or #len_bytes == 0 then
		break
	end
	assert(#len_bytes == 2, "truncated tcpdns stream: incomplete length prefix")
	local len = read_u16_be(len_bytes, 1)
	local msg = io.stdin:read(len)
	assert(msg and #msg == len, "truncated tcpdns stream")

	local packet = build_packet(msg)
	local rec = u32_be(ts_sec) .. u32_be(ts_usec) .. u32_be(#packet) .. u32_be(#packet)
	io.stdout:write(rec)
	io.stdout:write(packet)

	ts_usec = ts_usec + spacing_us
	while ts_usec >= 1000000 do
		ts_usec = ts_usec - 1000000
		ts_sec = ts_sec + 1
	end
	count = count + 1
end

io.stderr:write(string.format("wrote %d packets to stdout\n", count))

File: tests/tcpdns2pcap.py
#!/usr/bin/env python3
"""
Convert a raw DNS-TCP-framed stream (2-byte BE length prefix + DNS wire
message, repeated) on stdin into a PCAP file containing IPv6/UDP/DNS
packets on stdout.

Writes a classic (non-nanosecond) PCAP file directly (LINKTYPE_RAW=101,
raw IP packets, no link-layer header).

Always uses ::1 and port 53 on both ends.

usage: tcpdns2pcap.py [spacing_us] < in.tcpdns > out.pcap
"""

import argparse
import struct
import sys

SRC_PORT = 53
DST_PORT = 53

IP6_LOOPBACK = b"\x00" * 15 + b"\x01"  # ::1

LINKTYPE_RAW = 101  # tcpdump.org registry value, portable (not OS-local DLT_RAW)


def build_packet(dns_wire: bytes, client_id: int = 1) -> bytes:
    udp_len = 2 + 2 + 2 + 2 + len(dns_wire)

    udp_hdr = struct.pack(
        "!HHHH", SRC_PORT, DST_PORT, udp_len, 0
    )  # checksum 0 = unused

    ip6_hdr = (
        bytes([0x60, 0x00, 0x00, 0x00])  # version=6, traffic class/flow label=0
        + struct.pack("!H", udp_len)  # payload length
        + bytes([17])  # next header = UDP
        + bytes([64])  # hop limit
        + client_id.to_bytes(16, "big")  # source address; ipsplit's client key
        + IP6_LOOPBACK  # dest address; ipsplit rewrites this regardless
    )

    return ip6_hdr + udp_hdr + dns_wire


def pcap_global_header() -> bytes:
    # All-big-endian incl. magic -> readers detect swapped magic and
    # byte-swap the rest (standard "swapped-endian" pcap dialect).
    return struct.pack(
        "!IHHIIII",
        0xA1B2C3D4,  # magic
        2,  # version_major
        4,  # version_minor
        0,  # thiszone
        0,  # sigfigs
        65535,  # snaplen
        LINKTYPE_RAW,  # network
    )


def read_next_message(stdin) -> bytes | None:
    """One length-prefixed tcpdns message, or None on clean EOF."""
    len_bytes = stdin.read(2)
    if not len_bytes:
        return None
    assert len(len_bytes) == 2, "truncated tcpdns stream: incomplete length prefix"
    (length,) = struct.unpack("!H", len_bytes)
    msg = stdin.read(length)
    assert len(msg) == length, "truncated tcpdns stream"
    return msg


def write_packet_record(
    stdout, dns_wire: bytes, ts_sec: int, ts_usec: int, client_id: int = 1
) -> None:
    packet = build_packet(dns_wire, client_id)
    rec = struct.pack("!IIII", ts_sec, ts_usec, len(packet), len(packet))
    stdout.write(rec)
    stdout.write(packet)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "spacing_us",
        nargs="?",
        type=int,
        default=300000,
        help="spacing between packets in us (default: 300 000)",
    )
    spacing_us = parser.parse_args().spacing_us

    stdin = sys.stdin.buffer
    stdout = sys.stdout.buffer

    stdout.write(pcap_global_header())

    ts_sec = 0
    ts_usec = 0
    count = 0
    while True:
        msg = read_next_message(stdin)
        if msg is None:
            break

        write_packet_record(stdout, msg, ts_sec, ts_usec)

        ts_usec += spacing_us
        while ts_usec >= 1000000:
            ts_usec -= 1000000
            ts_sec += 1
        count += 1

    print(f"wrote {count} packets to stdout", file=sys.stderr)


if __name__ == "__main__":
    main()
